src/utils/appinsights.ts

```diff
@@ -85,3 +91,42 @@ export function removeQueryParamsPreprocessor(
   }
   return true;
 }
+
+/**
+ * If the user objects provides the session_tracking_id property, attach it to
+ * the current correlation context and propagate it to outgoing requests
+ */
+export function attachSessionTrackingId(user: User): void {
+  if (user.session_tracking_id !== undefined) {
+    // if we stored a session tracking ID in the user object, attach it to
+    // the current appinsights context and propagate it to the outgoing
+    // HTTP requests
+    appInsights
+      .getCorrelationContext()
+      .customProperties.setProperty(
```

Comment:
Are you sure this setting isn't applied to the shared defaultClient ? In case you probably want to look at
`runWithContext()` method, see
https://github.com/pagopa/io-functions-commons/blob/master/src/utils/application_insights.ts#L180

Reply:
`getCorrelationContext` returns the per-request context, see [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/156036ccc68cc250486260e556167ad06491361d/applicationinsights.ts#L127) - I'm actually unsure on how to access `customProperties` from the telemetry processor

Reply:
just checked the code, I needed to call `getProperty` to get the value of a custom property

Reply:
I'm not sure using the context this way is safe between subsequent calls (maybe a unit test with some inteleaved calls using setTimeout will clarify this)

Reply:
This is the recommended approach, see [this issue](https://github.com/microsoft/ApplicationInsights-node.js/issues/356#issuecomment-402243264)

`getCorrelationContext()` calls `getCurrentContext()` that returns [CorrelationContextManager.session.get(CorrelationContextManager.CONTEXT_NAME)](https://github.com/microsoft/ApplicationInsights-node.js/blob/156036ccc68cc250486260e556167ad06491361d/applicationinsights.ts#L141), the same value provided by [runWithContext](https://github.com/microsoft/ApplicationInsights-node.js/blob/5bf1848be22418b74e67a422f6daed651314c2a5/AutoCollection/CorrelationContextManager.ts#L101)

Reply:
Yes but `runWithContext` creates a new session with the provided context and only then it calls the wrapped method: the provided context - if not shared with the outer code (ie. generated) - cannot be mutated outside f() (of course if you pass as context the same vaule returned from `getCurrentContext()` you don't need `runWithContext`).

Said that, I'm not sure yet that the context you get from `getCurrentContext()` is not shared and immutable from outside the .

Reply:
update: probably I see now how the CorrelationContext may actually be per-request. it looks like it's using [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked/issues) that creates a new namespace for each request when called inside an express request handler

Reply:
I've found also this one: https://github.com/microsoft/ApplicationInsights-node.js/issues/392#issuecomment-387532917 (maybe we should add a comment that getCorrelationContext() is scoped to the express request handler)

Reply:
I've added an explanation and a link to the thread